### PR TITLE
Add python-six automated tests

### PR DESCRIPTION
**Pull request by skliu**
This module is a suite of tests for python-six, and outputs
the TAP format result.

- Related ticket: https://progress.opensuse.org/issues/52346
    Needles: no
    Verification run:
- SLE12SP2: http://10.67.19.78/tests/1280
- SLE12SP3: http://10.67.19.78/tests/1281
- SLE12SP4: http://10.67.19.78/tests/1282
- SLE12SP5: http://10.67.19.78/tests/1283
- SLE15 : http://10.67.19.78/tests/1279
- SLE15SP1: http://10.67.19.78/tests/1277

### CODE REVIEW

**Comment by skliu** [comment]
@foursixnine Could you please review my python-six testsuite? you've reviewed my "yast firewall" cmd automation testsuite, and suggested me using shell scripts instead of openqa api, this is a similar one. :)

File: tests/yast2_cmd/yast_ftp_server.pm [hunk 1]
```diff
@@ -110,7 +110,7 @@ sub run {
     if (is_sle('15+')) {
         assert_script_run 'yast ftp-server startup socket';
     }
-    if (is_sle('<=12-sp4')) {
+    if (is_sle('<15')) {
```
[the remark this answers is not in the document]
Comment (skliu) [reply tests/yast2_cmd/yast_ftp_server.pm]: yes, it's unrelated. It fixed https://progress.opensuse.org/issues/59028, should I open a stand-alone PR?

**Comment by skliu** [comment]
> It's looking very good! My only concern about tappy is that since it's on a third party source, it could fail, and the test would fail in turn too... I think packaging it would be better?

I've downloaded tappy to data/

File: tests/console/python_six.pm [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        assert_script_run('pip2 install tap.py; pip3 install tap.py');
+    } else {
+        # tappy cannot be installed using easy_install on sle12(compile error appears randomly), so use pip.
+        zypper_call('install python3');
+        assert_script_run('curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py');
+        assert_script_run('python2 get-pip.py; python3 get-pip.py');
+        assert_script_run('pip2 install tap.py; pip3 install tap.py');
+        assert_script_run('rm -rf get-pip.py');
+    }
+
```
[the remark this answers is not in the document]
Comment (skliu) [reply tests/console/python_six.pm]: fixed.

File: tests/console/python_six.pm [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+sub run {
+    select_console 'root-console';
+    zypper_call('install python-six python3-six');
+
+    if (is_sle('15+')) {
+        zypper_call('install python2-pip python3-pip');
+        assert_script_run('pip2 install tap.py; pip3 install tap.py');
+    } else {
+        # tappy cannot be installed using easy_install on sle12(compile error appears randomly), so use pip.
```
[the remark this answers is not in the document]
Comment (skliu) [reply tests/console/python_six.pm]: fixed.

File: tests/console/python_six.pm [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+# - In SUT, run python_six.py, and produce the result to a TAP format file.
+
+use base 'consoletest';
+use strict;
+use testapi;
+use utils;
+use version_utils 'is_sle';
+
+sub run_test{
+    my $my_py = $_[0];
```
[the remark this answers is not in the document]
Comment (skliu) [reply tests/console/python_six.pm]: fixed.

File: data/python_six.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        self.assertNotIsInstance(X(), six.class_types)
+
+###########################
+#category3 Object model compatibility
+###########################
+class test_object_model(unittest.TestCase):
+    #1 Unbound (class) method objects: no self
+    # Accessing a function attribute of a class by qualifying the class returns an unbound method object.
+    #2 Bound (instance) method objects:
+    #Accessing a function attribute of a class by qualifying an instance returns a bound method object.
```
[the remark this answers is not in the document]
Comment (skliu) [reply data/python_six.py]: fixed.

File: data/python_six.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+            pass
+        class Y(object):
+            pass
+        self.assertIsInstance(X, six.class_types)
+        self.assertIsInstance(Y, six.class_types)
+        self.assertNotIsInstance(X(), six.class_types)
+
+###########################
+#category3 Object model compatibility
+###########################
```
[the remark this answers is not in the document]
Comment (skliu) [reply data/python_six.py]: fixed.

File: data/python_six.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
+# -*- coding: utf-8 -*-
+import sys, unittest, types
+import six, tap
+
+###########################
+#Six provides simple utilities for wrapping over differences between Python 2 and Python 3. It is intended to support codebases that work on both Python 2 and 3 without modification.
+
+#Six wraps over differences in roughly 5 categorys:
+#category1 python2 and python3's versions and constants.
+###########################
```
[the remark this answers is not in the document]
Comment (skliu) [reply data/python_six.py]: fixed.

File: data/python_six.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+        def f(self):
+            return self
+        u = six.create_unbound_method(f, X)
+        if six.PY2:
+            self.assertIsInstance(u, types.MethodType)
+        x = X()
+        self.assertIs(f(x), x)
+
+    # six.create_bound_method(func, obj)
+    # Return a method object wrapping func and bound to obj. On both Python 2 and 3, this will return a types.MethodType object. The reason this wrapper exists is that on Python 2, the MethodType constructor requires the obj’s class to be passed.
```
[the remark this answers is not in the document]
Comment (skliu) [reply data/python_six.py]: fixed and chopped the length of a line to maximum 120 characters.

**Comment by foursixnine** [comment]
> > It's looking very good! My only concern about tappy is that since it's on a third party source, it could fail, and the test would fail in turn too... I think packaging it would be better?
> 
> I've downloaded tappy and packaged it in data/python_six/, so we can use it independently.

You're adding pip to the repository?. I'm sorry but that's a no go. Use the pip from SLE, or again, package tappy

**Comment by skliu** [comment]
> @skliu I appreciate the effort but, you're adding pip to the repository?. I'm sorry but that's a no go. Use the pip from SLE, If there's a bug in pip for sle, add it as a comment too, or maybe package tappy in obs?

@foursixnine Hi, there is pip rpm package on sle15 but not  on sle12. I will package pip(only on sle12) and tappy in rpm format and add them to QA Head, is this OK?

**Comment by foursixnine** [comment]
> > @skliu I appreciate the effort but, you're adding pip to the repository?. I'm sorry but that's a no go. Use the pip from SLE, If there's a bug in pip for sle, add it as a comment too, or maybe package tappy in obs?
> 
> @foursixnine Hi, there is pip rpm package on sle15 but not on sle12. I will package pip(only on sle12) and tappy in rpm format and add them to QA Head, is this OK?

If you package tappy, pip will not be needed :) This is fine for me

File: tests/yast2_cmd/yast_ftp_server.pm (hunk 1 above)
[the remark this answers is not in the document]
Comment (mimi1vx) [reply tests/yast2_cmd/yast_ftp_server.pm]: yes please

**Comment by mimi1vx** [comment]
please package tappy to devel:languages:python and submit it for tumbleweed and as MaintIncident to Leap 15.1/2

**Comment by mimi1vx** [comment]
btw `six` package has own test suite ( and running in OBS as python-six:test )

**Comment by skliu** [comment]
> please package tappy to devel:languages:python and submit it for tumbleweed and as MaintIncident to Leap 15.1/2
there is python-tap.py on ibs. I've built it on sle. please review my code.

**Comment by plusky** [comment]
https://build.suse.de/package/rdiff/home:shukui/python-tap.py?opackage=python-tap.py&oproject=SUSE%3AFactory%3AHead&rev=4 - this is very very wrong, please make sure that package builds on supported distros, dont just remove dependencies and disable tests of package itself. And submit changes to devel project in OBS.

**Comment by skliu** [comment]
> https://build.suse.de/package/rdiff/home:shukui/python-tap.py?opackage=python-tap.py&oproject=SUSE%3AFactory%3AHead&rev=4 - this is very very wrong, please make sure that package builds on supported distros, dont just remove dependencies and disable tests of package itself. And submit changes to devel project in OBS.

Hi Martin, there is python-tap.py in devel:languages:python and devel:languages:python:backports on build.opensuse.org. Do you mean that I should branch them, keep the dependencies and %check section in python-tap.py.spec, make them be built successfully on all my test platforms(sle12sp2, sle12sp3, sle12sp4, sle12sp5, sle15, sle15sp1), and then submit changes back to devel:languages:python and devel:languages:python:backports in OBS?

**Comment by plusky** [comment]
I have linked python-tap.py to QA:Head - see https://build.suse.de/package/show/QA:Head/python-tap.py but it has satisfied dependencies only on SLE-15

File: tests/console/python_six.pm [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    select_console 'root-console';
+    add_qa_head_repo;
+
+    if (is_sle('15+')) {
+        zypper_call('install python2-tap.py python3-tap.py');
+    } else {
+        zypper_call('install python-tap.py python3-tap.py');
+    }
+
+    zypper_call('install python-six python3-six python-tap.py python3-tap.py');
```
Comment (foursixnine) [inline tests/console/python_six.pm]:
```suggestion
    zypper_call('install python-six python3-six');
```

File: tests/console/python_six.pm [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    my $python_six_dir = "/home/bernhard/data/python_six";
+    if (script_run("test -d $python_six_dir")) {    #dir does not exist.
+        my $archive = "python_six.data";
+        assert_script_run(
+            sprintf "cd; curl -L -v %s/data/python_six > %s && cpio -id < %s && mv data python_six && ls python_six",
```
Comment (foursixnine) [inline tests/console/python_six.pm]: Please don't use sprintf, take a look here: https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/8839#discussion_r344180881

**Comment by foursixnine** [comment]
@skliu I think the following files can be removed
```
	new file:   data/python_six.py
	new file:   data/python_six/get-pip.py
	new file:   data/python_six/python_six.py
	new file:   data/python_six/tap.py-2.6.2-py2.py3-none-any.whl
```

Also please write a better commit message/title: 

E.g: "Add tests for python-six", give it a read to: https://chris.beams.io/posts/git-commit/

File: tests/console/python_six.pm (hunk 9 above)
Comment (skliu) [reply tests/console/python_six.pm]: fixed.

File: tests/console/python_six.pm (hunk 10 above)
Comment (skliu) [reply tests/console/python_six.pm]: fixed.

**Comment by skliu** [comment]
> @skliu I think the following files can be removed
> 
> ```
> 	new file:   data/python_six.py
> 	new file:   data/python_six/get-pip.py
> 	new file:   data/python_six/python_six.py
> 	new file:   data/python_six/tap.py-2.6.2-py2.py3-none-any.whl
> ```
> 
> Also please write a better commit message/title:
> 
> E.g: "Add tests for python-six", give it a read to: https://chris.beams.io/posts/git-commit/

fixed.

**Comment by skliu** [comment]
> @skliu ~can you provide verification runs with the new code?~ Only 12 sp5 is failing now, but I guess it's due to the product not being released yet

I've run the new code using openqa and updated the testing urls in the first comment, they are all failed except sle15sp1.

**Comment by skliu** [comment]
> I think the package needs to be submitted to the other codestreams

Is there anything else I can do? I think I can write an email to the author of the package python-tap.py:  tchvatal@suse.com

**Comment by foursixnine** [comment]
@skliu Do you plan on working on this pr again?

**Comment by skliu** [comment]
> @skliu Do you plan on working on this pr again?

Yes. I've added devel:languages:python:backports to my code, then the tests passed: http://10.67.19.78/tests/1303

File: tests/console/python_six.pm [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+sub run_test {
+    my ($my_py)     = @_;
+    my $tap_results = "results.tap";
+    my $ret         = script_run("prove -ve $my_py python_six.py >> $tap_results");
+    return $ret;
+}
+
+sub run {
+    select_console 'root-console';
```
Comment (foursixnine) [inline tests/console/python_six.pm:34]: use serial console here

File: tests/console/python_six.pm [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    add_qa_head_repo;
+
+    if (is_sle('15+')) {
+        zypper_call('install python2-tap.py python3-tap.py');
+    } else {
+        zypper_call('install python-tap.py python3-tap.py');
+    }
+
+    zypper_call('install python-six python3-six');
```
Comment (foursixnine) [inline tests/console/python_six.pm:43]: I fail to see how the test would pass when python-six is not available in the product itself (https://scc.suse.com/packages?name=SUSE%20Linux%20Enterprise%20Server&version=15.1&arch=x86_64&query=python-six&module=) for already released packages. I guess that's already a bug? or is intented to test it from a different repository?

Comment (skliu) [reply tests/console/python_six.pm:43]: Hi, since there is testsuite within the python-six package,  I think this automated tests is not necessary.

**Pull request closed without merging**